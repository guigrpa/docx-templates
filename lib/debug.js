"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setDebugLogSink = exports.logger = void 0;
exports.logger = { debug: function () { } };
function setDebugLogSink(f) {
    exports.logger.debug = f;
}
exports.setDebugLogSink = setDebugLogSink;
