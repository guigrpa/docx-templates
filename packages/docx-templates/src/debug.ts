// Cannot be covered by Flow! The purpose of this file is to isolate
// an optional dependency, only used in development (so that Flow
// doesn't throw errors when processing users' projects)
import { mainStory, chalk, addListener } from 'storyboard';
import consoleListener from 'storyboard-listener-console';

addListener(consoleListener);

export { mainStory, chalk };
