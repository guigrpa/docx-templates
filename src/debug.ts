export default { info: console.log, debug: console.log };
