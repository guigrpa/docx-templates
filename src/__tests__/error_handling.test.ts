import path from 'path';
import fs from 'fs';
import QR from 'qrcode';
import { createReport } from '../index';
import { setDebugLogSink } from '../debug';
import {
  NullishCommandResultError,
  CommandExecutionError,
  InvalidCommandError,
} from '../errors';

if (process.env.DEBUG) setDebugLogSink(console.log);

['noSandbox', 'sandbox'].forEach(sbStatus => {
  const noSandbox = sbStatus === 'sandbox' ? false : true;

  describe(`${sbStatus}`, () => {
    describe('rejectNullish setting', () => {
      it('INS', async () => {
        const template = await fs.promises.readFile(
          path.join(__dirname, 'fixtures', 'rejectNullishINS.docx')
        );

        // When not explicitly set, rejectNullish should be considered 'false' so this case should resolve.
        await expect(
          createReport({
            noSandbox,
            template,
            data: {
              testobj: {}, // accessing a non-existing property will result in `undefined`
              test2: 'second value!',
            },
          })
        ).resolves.toBeInstanceOf(Uint8Array);

        // The same case should throw when we decide NOT to accept nullish values.
        await expect(
          createReport({
            noSandbox,
            template,
            data: {
              testobj: {}, // accessing a non-existing property will result in `undefined`
              test2: 'second value!',
            },
            rejectNullish: true,
          })
        ).rejects.toBeInstanceOf(Error);

        // Should be ok when we actually set the value.
        await expect(
          createReport({
            noSandbox,
            template,
            data: {
              testobj: { value: 'the value is now set' },
              test2: 'second value!',
            },
            rejectNullish: true,
          })
        ).resolves.toBeInstanceOf(Uint8Array);
      });

      it('IMAGE', async () => {
        const template = await fs.promises.readFile(
          path.join(__dirname, 'fixtures', 'rejectNullishIMAGE.docx')
        );
        await expect(
          createReport({
            noSandbox,
            template,
            data: {},
            additionalJsContext: {
              qr: () => undefined,
            },
          })
        ).resolves.toBeInstanceOf(Uint8Array);

        await expect(
          createReport({
            noSandbox,
            template,
            data: {},
            rejectNullish: true,
            additionalJsContext: {
              qr: () => undefined,
            },
          })
        ).rejects.toThrowErrorMatchingSnapshot();

        await expect(
          createReport({
            noSandbox,
            template,
            data: {},
            rejectNullish: true,
            additionalJsContext: {
              qr: async (contents: string) => {
                const dataUrl = await QR.toDataURL(contents, { width: 500 });
                const data = dataUrl.slice('data:image/gif;base64,'.length);
                return { width: 6, height: 6, data, extension: '.gif' };
              },
            },
          })
        ).resolves.toBeInstanceOf(Uint8Array);
      });
    });

    describe('custom ErrorHandler', () => {
      it('allows graceful handling of NullishCommandResultError', async () => {
        expect.assertions(3);

        const template = await fs.promises.readFile(
          path.join(__dirname, 'fixtures', 'rejectNullishINS.docx')
        );

        const result = await createReport(
          {
            noSandbox,
            template,
            data: {
              testobj: {}, // accessing a non-existing property will result in `undefined`
              test2: 'second value!',
            },
            rejectNullish: true,
            errorHandler: (err, code) => {
              expect(err).toBeInstanceOf(NullishCommandResultError);
              expect(code).toStrictEqual('testobj.value');
              return 'its ok buddy';
            },
          },
          'XML'
        );
        expect(result).toMatchSnapshot();
      });

      it('handles arbitrary errors occurring in command execution', async () => {
        const template = await fs.promises.readFile(
          path.join(__dirname, 'fixtures', 'commandExecutionError.docx')
        );

        // First check whether the CommandExecutionError is triggered correctly
        await expect(
          createReport({ noSandbox, template, data: {} })
        ).rejects.toThrowError(CommandExecutionError);

        // Now try with an errorHandler
        expect(
          await createReport(
            {
              noSandbox,
              template,
              data: {},
              errorHandler: (err, code) => 'no problem dude',
            },
            'XML'
          )
        ).toMatchSnapshot();
      });

      it('properly handles InvalidCommandError', async () => {
        const template = await fs.promises.readFile(
          path.join(__dirname, 'fixtures', 'invalidMultipleErrors.docx')
        );

        const errs: Error[] = [];
        expect(
          await createReport(
            {
              noSandbox,
              template,
              data: {
                companies: [
                  { name: 'FIRST' },
                  { name: 'SECOND' },
                  { name: 'THIRD' },
                ],
              },
              errorHandler: (err, code) => {
                errs.push(err);
                return 'its ok buddy';
              },
            },
            'XML'
          )
        ).toMatchSnapshot();

        expect(errs.some(e => e instanceof InvalidCommandError)).toBeTruthy();
      });

      it('handler can decide to re-throw the error, crashing the render', async () => {
        const template = await fs.promises.readFile(
          path.join(__dirname, 'fixtures', 'invalidMultipleErrors.docx')
        );

        await expect(
          createReport({
            noSandbox,
            template,
            data: {
              companies: [
                { name: 'FIRST' },
                { name: 'SECOND' },
                { name: 'THIRD' },
              ],
            },
            errorHandler: (err, code) => {
              throw new Error('yeah, no!');
            },
          })
        ).rejects.toThrowError('yeah, no!');
      });
    });
  });

  it('throw when user tries to iterate over non-array', async () => {
    const template = await fs.promises.readFile(
      path.join(__dirname, 'fixtures', 'forOverObject.docx')
    );
    await expect(
      createReport({
        noSandbox,
        template,
        data: {
          companies: {
            one: 'FIRST',
            two: 'SECOND',
            three: 'THIRD',
          },
        },
      })
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it('throw when result of INS command is an object', async () => {
    const template = await fs.promises.readFile(
      path.join(__dirname, 'fixtures', 'objectCommandResultError.docx')
    );
    await expect(
      createReport({
        noSandbox,
        template,
        data: {
          companies: {
            one: 'FIRST',
            two: 'SECOND',
            three: 'THIRD',
          },
        },
      })
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});
