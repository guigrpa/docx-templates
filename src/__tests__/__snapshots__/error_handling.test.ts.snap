// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Incomplete conditional statement: missing END-IF 1`] = `
[
  [Error: Incomplete IF/END-IF statement. Make sure each IF-statement has a corresponding END-IF command.],
]
`;

exports[`Incomplete conditional statement: missing END-IF 2`] = `
[
  [Error: Incomplete IF/END-IF statement. Make sure each IF-statement has a corresponding END-IF command.],
]
`;

exports[`noSandbox custom ErrorHandler allows graceful handling of NullishCommandResultError 1`] = `
"<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<w:document xmlns:wpc="http://schemas.microsoft.com/office/word/2010/wordprocessingCanvas" xmlns:cx="http://schemas.microsoft.com/office/drawing/2014/chartex" xmlns:cx1="http://schemas.microsoft.com/office/drawing/2015/9/8/chartex" xmlns:cx2="http://schemas.microsoft.com/office/drawing/2015/10/21/chartex" xmlns:cx3="http://schemas.microsoft.com/office/drawing/2016/5/9/chartex" xmlns:cx4="http://schemas.microsoft.com/office/drawing/2016/5/10/chartex" xmlns:cx5="http://schemas.microsoft.com/office/drawing/2016/5/11/chartex" xmlns:cx6="http://schemas.microsoft.com/office/drawing/2016/5/12/chartex" xmlns:cx7="http://schemas.microsoft.com/office/drawing/2016/5/13/chartex" xmlns:cx8="http://schemas.microsoft.com/office/drawing/2016/5/14/chartex" xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006" xmlns:aink="http://schemas.microsoft.com/office/drawing/2016/ink" xmlns:am3d="http://schemas.microsoft.com/office/drawing/2017/model3d" xmlns:o="urn:schemas-microsoft-com:office:office" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships" xmlns:m="http://schemas.openxmlformats.org/officeDocument/2006/math" xmlns:v="urn:schemas-microsoft-com:vml" xmlns:wp14="http://schemas.microsoft.com/office/word/2010/wordprocessingDrawing" xmlns:wp="http://schemas.openxmlformats.org/drawingml/2006/wordprocessingDrawing" xmlns:w10="urn:schemas-microsoft-com:office:word" xmlns:w="http://schemas.openxmlformats.org/wordprocessingml/2006/main" xmlns:w14="http://schemas.microsoft.com/office/word/2010/wordml" xmlns:w15="http://schemas.microsoft.com/office/word/2012/wordml" xmlns:w16cid="http://schemas.microsoft.com/office/word/2016/wordml/cid" xmlns:w16se="http://schemas.microsoft.com/office/word/2015/wordml/symex" xmlns:wpg="http://schemas.microsoft.com/office/word/2010/wordprocessingGroup" xmlns:wpi="http://schemas.microsoft.com/office/word/2010/wordprocessingInk" xmlns:wne="http://schemas.microsoft.com/office/word/2006/wordml" xmlns:wps="http://schemas.microsoft.com/office/word/2010/wordprocessingShape" mc:Ignorable="w14 w15 w16se w16cid wp14">
  <w:body>
    <w:p w14:paraId="3F0B5C58" w14:textId="6F568B18" w:rsidR="00F35C33" w:rsidRDefault="00D81685">
      <w:pPr>
        <w:rPr>
          <w:lang w:val="nl-NL"/>
        </w:rPr>
      </w:pPr>
      <w:r>
        <w:rPr>
          <w:lang w:val="nl-NL"/>
        </w:rPr>
        <w:t xml:space="preserve">Error: Result of command testobj.value is null or undefined and rejectNullish is set</w:t>
      </w:r>
      <w:bookmarkStart w:id="0" w:name="_GoBack"/>
      <w:bookmarkEnd w:id="0"/>
      <w:r>
        <w:rPr>
          <w:lang w:val="nl-NL"/>
        </w:rPr>
        <w:t xml:space="preserve"></w:t>
      </w:r>
    </w:p>
    <w:p w14:paraId="5AFBB39A" w14:textId="7977431C" w:rsidR="00D81685" w:rsidRPr="00D81685" w:rsidRDefault="00D81685" w:rsidP="00D81685">
      <w:pPr>
        <w:rPr>
          <w:lang w:val="nl-NL"/>
        </w:rPr>
      </w:pPr>
      <w:r>
        <w:rPr>
          <w:lang w:val="nl-NL"/>
        </w:rPr>
        <w:t xml:space="preserve">second value!</w:t>
      </w:r>
      <w:r>
        <w:rPr>
          <w:lang w:val="nl-NL"/>
        </w:rPr>
        <w:t xml:space="preserve"></w:t>
      </w:r>
      <w:r>
        <w:rPr>
          <w:lang w:val="nl-NL"/>
        </w:rPr>
        <w:t xml:space="preserve"></w:t>
      </w:r>
    </w:p>
    <w:p w14:paraId="3719270B" w14:textId="77777777" w:rsidR="00D81685" w:rsidRPr="00D81685" w:rsidRDefault="00D81685">
      <w:pPr>
        <w:rPr>
          <w:lang w:val="nl-NL"/>
        </w:rPr>
      </w:pPr>
    </w:p>
    <w:sectPr w:rsidR="00D81685" w:rsidRPr="00D81685" w:rsidSect="0019199C">
      <w:pgSz w:w="11900" w:h="16840"/>
      <w:pgMar w:top="1440" w:right="1440" w:bottom="1440" w:left="1440" w:header="708" w:footer="708" w:gutter="0"/>
      <w:cols w:space="708"/>
      <w:docGrid w:linePitch="360"/>
    </w:sectPr>
  </w:body>
</w:document>"
`;

exports[`noSandbox custom ErrorHandler handles arbitrary errors occurring in command execution 1`] = `
"<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<w:document xmlns:wpc="http://schemas.microsoft.com/office/word/2010/wordprocessingCanvas" xmlns:cx="http://schemas.microsoft.com/office/drawing/2014/chartex" xmlns:cx1="http://schemas.microsoft.com/office/drawing/2015/9/8/chartex" xmlns:cx2="http://schemas.microsoft.com/office/drawing/2015/10/21/chartex" xmlns:cx3="http://schemas.microsoft.com/office/drawing/2016/5/9/chartex" xmlns:cx4="http://schemas.microsoft.com/office/drawing/2016/5/10/chartex" xmlns:cx5="http://schemas.microsoft.com/office/drawing/2016/5/11/chartex" xmlns:cx6="http://schemas.microsoft.com/office/drawing/2016/5/12/chartex" xmlns:cx7="http://schemas.microsoft.com/office/drawing/2016/5/13/chartex" xmlns:cx8="http://schemas.microsoft.com/office/drawing/2016/5/14/chartex" xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006" xmlns:aink="http://schemas.microsoft.com/office/drawing/2016/ink" xmlns:am3d="http://schemas.microsoft.com/office/drawing/2017/model3d" xmlns:o="urn:schemas-microsoft-com:office:office" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships" xmlns:m="http://schemas.openxmlformats.org/officeDocument/2006/math" xmlns:v="urn:schemas-microsoft-com:vml" xmlns:wp14="http://schemas.microsoft.com/office/word/2010/wordprocessingDrawing" xmlns:wp="http://schemas.openxmlformats.org/drawingml/2006/wordprocessingDrawing" xmlns:w10="urn:schemas-microsoft-com:office:word" xmlns:w="http://schemas.openxmlformats.org/wordprocessingml/2006/main" xmlns:w14="http://schemas.microsoft.com/office/word/2010/wordml" xmlns:w15="http://schemas.microsoft.com/office/word/2012/wordml" xmlns:w16cid="http://schemas.microsoft.com/office/word/2016/wordml/cid" xmlns:w16se="http://schemas.microsoft.com/office/word/2015/wordml/symex" xmlns:wpg="http://schemas.microsoft.com/office/word/2010/wordprocessingGroup" xmlns:wpi="http://schemas.microsoft.com/office/word/2010/wordprocessingInk" xmlns:wne="http://schemas.microsoft.com/office/word/2006/wordml" xmlns:wps="http://schemas.microsoft.com/office/word/2010/wordprocessingShape" mc:Ignorable="w14 w15 w16se w16cid wp14">
  <w:body>
    <w:p w14:paraId="52DAE836" w14:textId="35F10B17" w:rsidR="00A954E6" w:rsidRPr="00537D95" w:rsidRDefault="00A954E6" w:rsidP="00A954E6">
      <w:bookmarkStart w:id="0" w:name="_GoBack"/>
      <w:r w:rsidRPr="00537D95">
        <w:t xml:space="preserve">no problem dude</w:t>
      </w:r>
      <w:r>
        <w:t xml:space="preserve"></w:t>
      </w:r>
      <w:r w:rsidRPr="00537D95">
        <w:t xml:space="preserve"></w:t>
      </w:r>
    </w:p>
    <w:bookmarkEnd w:id="0"/>
    <w:p w14:paraId="35401385" w14:textId="7B1795CE" w:rsidR="00413949" w:rsidRPr="00A954E6" w:rsidRDefault="00413949" w:rsidP="00A954E6"/>
    <w:sectPr w:rsidR="00413949" w:rsidRPr="00A954E6">
      <w:pgSz w:w="12240" w:h="15840"/>
      <w:pgMar w:top="1417" w:right="1701" w:bottom="1417" w:left="1701" w:header="708" w:footer="708" w:gutter="0"/>
      <w:cols w:space="708"/>
      <w:docGrid w:linePitch="360"/>
    </w:sectPr>
  </w:body>
</w:document>"
`;

exports[`noSandbox custom ErrorHandler properly handles InvalidCommandError 1`] = `
"<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<w:document xmlns:wpc="http://schemas.microsoft.com/office/word/2010/wordprocessingCanvas" xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006" xmlns:o="urn:schemas-microsoft-com:office:office" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships" xmlns:m="http://schemas.openxmlformats.org/officeDocument/2006/math" xmlns:v="urn:schemas-microsoft-com:vml" xmlns:wp14="http://schemas.microsoft.com/office/word/2010/wordprocessingDrawing" xmlns:wp="http://schemas.openxmlformats.org/drawingml/2006/wordprocessingDrawing" xmlns:w10="urn:schemas-microsoft-com:office:word" xmlns:w="http://schemas.openxmlformats.org/wordprocessingml/2006/main" xmlns:w14="http://schemas.microsoft.com/office/word/2010/wordml" xmlns:w15="http://schemas.microsoft.com/office/word/2012/wordml" xmlns:wpg="http://schemas.microsoft.com/office/word/2010/wordprocessingGroup" xmlns:wpi="http://schemas.microsoft.com/office/word/2010/wordprocessingInk" xmlns:wne="http://schemas.microsoft.com/office/word/2006/wordml" xmlns:wps="http://schemas.microsoft.com/office/word/2010/wordprocessingShape" mc:Ignorable="w14 w15 wp14">
  <w:body>
    <w:p w:rsidR="00A17023" w:rsidRPr="00537D95" w:rsidRDefault="00A17023">
      <w:r w:rsidRPr="00537D95">
        <w:t xml:space="preserve">Here is the query…</w:t>
      </w:r>
    </w:p>
    <w:p w:rsidR="00413949" w:rsidRDefault="00413949">
      <w:r w:rsidRPr="0070400C">
        <w:t xml:space="preserve">Invalid command…</w:t>
      </w:r>
    </w:p>
    <w:p w:rsidR="00537D95" w:rsidRPr="0070400C" w:rsidRDefault="00537D95">
      <w:r>
        <w:t xml:space="preserve">SyntaxError: Unexpected identifier 'foo'</w:t>
      </w:r>
    </w:p>
    <w:p w:rsidR="00413949" w:rsidRPr="0070400C" w:rsidRDefault="00413949">
      <w:bookmarkStart w:id="0" w:name="_GoBack"/>
      <w:bookmarkEnd w:id="0"/>
      <w:r w:rsidRPr="0070400C">
        <w:t xml:space="preserve">Post command…</w:t>
      </w:r>
    </w:p>
    <w:sectPr w:rsidR="00413949" w:rsidRPr="0070400C">
      <w:pgSz w:w="12240" w:h="15840"/>
      <w:pgMar w:top="1417" w:right="1701" w:bottom="1417" w:left="1701" w:header="708" w:footer="708" w:gutter="0"/>
      <w:cols w:space="708"/>
      <w:docGrid w:linePitch="360"/>
    </w:sectPr>
  </w:body>
</w:document>"
`;

exports[`noSandbox custom ErrorHandler properly handles InvalidCommandError 2`] = `
[
  [SyntaxError: Unexpected identifier 'foo'],
]
`;

exports[`noSandbox rejectNullish setting IMAGE 1`] = `"Result of command qr(\`some random string\`) is null or undefined and rejectNullish is set"`;

exports[`sandbox custom ErrorHandler allows graceful handling of NullishCommandResultError 1`] = `
"<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<w:document xmlns:wpc="http://schemas.microsoft.com/office/word/2010/wordprocessingCanvas" xmlns:cx="http://schemas.microsoft.com/office/drawing/2014/chartex" xmlns:cx1="http://schemas.microsoft.com/office/drawing/2015/9/8/chartex" xmlns:cx2="http://schemas.microsoft.com/office/drawing/2015/10/21/chartex" xmlns:cx3="http://schemas.microsoft.com/office/drawing/2016/5/9/chartex" xmlns:cx4="http://schemas.microsoft.com/office/drawing/2016/5/10/chartex" xmlns:cx5="http://schemas.microsoft.com/office/drawing/2016/5/11/chartex" xmlns:cx6="http://schemas.microsoft.com/office/drawing/2016/5/12/chartex" xmlns:cx7="http://schemas.microsoft.com/office/drawing/2016/5/13/chartex" xmlns:cx8="http://schemas.microsoft.com/office/drawing/2016/5/14/chartex" xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006" xmlns:aink="http://schemas.microsoft.com/office/drawing/2016/ink" xmlns:am3d="http://schemas.microsoft.com/office/drawing/2017/model3d" xmlns:o="urn:schemas-microsoft-com:office:office" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships" xmlns:m="http://schemas.openxmlformats.org/officeDocument/2006/math" xmlns:v="urn:schemas-microsoft-com:vml" xmlns:wp14="http://schemas.microsoft.com/office/word/2010/wordprocessingDrawing" xmlns:wp="http://schemas.openxmlformats.org/drawingml/2006/wordprocessingDrawing" xmlns:w10="urn:schemas-microsoft-com:office:word" xmlns:w="http://schemas.openxmlformats.org/wordprocessingml/2006/main" xmlns:w14="http://schemas.microsoft.com/office/word/2010/wordml" xmlns:w15="http://schemas.microsoft.com/office/word/2012/wordml" xmlns:w16cid="http://schemas.microsoft.com/office/word/2016/wordml/cid" xmlns:w16se="http://schemas.microsoft.com/office/word/2015/wordml/symex" xmlns:wpg="http://schemas.microsoft.com/office/word/2010/wordprocessingGroup" xmlns:wpi="http://schemas.microsoft.com/office/word/2010/wordprocessingInk" xmlns:wne="http://schemas.microsoft.com/office/word/2006/wordml" xmlns:wps="http://schemas.microsoft.com/office/word/2010/wordprocessingShape" mc:Ignorable="w14 w15 w16se w16cid wp14">
  <w:body>
    <w:p w14:paraId="3F0B5C58" w14:textId="6F568B18" w:rsidR="00F35C33" w:rsidRDefault="00D81685">
      <w:pPr>
        <w:rPr>
          <w:lang w:val="nl-NL"/>
        </w:rPr>
      </w:pPr>
      <w:r>
        <w:rPr>
          <w:lang w:val="nl-NL"/>
        </w:rPr>
        <w:t xml:space="preserve">Error: Result of command testobj.value is null or undefined and rejectNullish is set</w:t>
      </w:r>
      <w:bookmarkStart w:id="0" w:name="_GoBack"/>
      <w:bookmarkEnd w:id="0"/>
      <w:r>
        <w:rPr>
          <w:lang w:val="nl-NL"/>
        </w:rPr>
        <w:t xml:space="preserve"></w:t>
      </w:r>
    </w:p>
    <w:p w14:paraId="5AFBB39A" w14:textId="7977431C" w:rsidR="00D81685" w:rsidRPr="00D81685" w:rsidRDefault="00D81685" w:rsidP="00D81685">
      <w:pPr>
        <w:rPr>
          <w:lang w:val="nl-NL"/>
        </w:rPr>
      </w:pPr>
      <w:r>
        <w:rPr>
          <w:lang w:val="nl-NL"/>
        </w:rPr>
        <w:t xml:space="preserve">second value!</w:t>
      </w:r>
      <w:r>
        <w:rPr>
          <w:lang w:val="nl-NL"/>
        </w:rPr>
        <w:t xml:space="preserve"></w:t>
      </w:r>
      <w:r>
        <w:rPr>
          <w:lang w:val="nl-NL"/>
        </w:rPr>
        <w:t xml:space="preserve"></w:t>
      </w:r>
    </w:p>
    <w:p w14:paraId="3719270B" w14:textId="77777777" w:rsidR="00D81685" w:rsidRPr="00D81685" w:rsidRDefault="00D81685">
      <w:pPr>
        <w:rPr>
          <w:lang w:val="nl-NL"/>
        </w:rPr>
      </w:pPr>
    </w:p>
    <w:sectPr w:rsidR="00D81685" w:rsidRPr="00D81685" w:rsidSect="0019199C">
      <w:pgSz w:w="11900" w:h="16840"/>
      <w:pgMar w:top="1440" w:right="1440" w:bottom="1440" w:left="1440" w:header="708" w:footer="708" w:gutter="0"/>
      <w:cols w:space="708"/>
      <w:docGrid w:linePitch="360"/>
    </w:sectPr>
  </w:body>
</w:document>"
`;

exports[`sandbox custom ErrorHandler handles arbitrary errors occurring in command execution 1`] = `
"<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<w:document xmlns:wpc="http://schemas.microsoft.com/office/word/2010/wordprocessingCanvas" xmlns:cx="http://schemas.microsoft.com/office/drawing/2014/chartex" xmlns:cx1="http://schemas.microsoft.com/office/drawing/2015/9/8/chartex" xmlns:cx2="http://schemas.microsoft.com/office/drawing/2015/10/21/chartex" xmlns:cx3="http://schemas.microsoft.com/office/drawing/2016/5/9/chartex" xmlns:cx4="http://schemas.microsoft.com/office/drawing/2016/5/10/chartex" xmlns:cx5="http://schemas.microsoft.com/office/drawing/2016/5/11/chartex" xmlns:cx6="http://schemas.microsoft.com/office/drawing/2016/5/12/chartex" xmlns:cx7="http://schemas.microsoft.com/office/drawing/2016/5/13/chartex" xmlns:cx8="http://schemas.microsoft.com/office/drawing/2016/5/14/chartex" xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006" xmlns:aink="http://schemas.microsoft.com/office/drawing/2016/ink" xmlns:am3d="http://schemas.microsoft.com/office/drawing/2017/model3d" xmlns:o="urn:schemas-microsoft-com:office:office" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships" xmlns:m="http://schemas.openxmlformats.org/officeDocument/2006/math" xmlns:v="urn:schemas-microsoft-com:vml" xmlns:wp14="http://schemas.microsoft.com/office/word/2010/wordprocessingDrawing" xmlns:wp="http://schemas.openxmlformats.org/drawingml/2006/wordprocessingDrawing" xmlns:w10="urn:schemas-microsoft-com:office:word" xmlns:w="http://schemas.openxmlformats.org/wordprocessingml/2006/main" xmlns:w14="http://schemas.microsoft.com/office/word/2010/wordml" xmlns:w15="http://schemas.microsoft.com/office/word/2012/wordml" xmlns:w16cid="http://schemas.microsoft.com/office/word/2016/wordml/cid" xmlns:w16se="http://schemas.microsoft.com/office/word/2015/wordml/symex" xmlns:wpg="http://schemas.microsoft.com/office/word/2010/wordprocessingGroup" xmlns:wpi="http://schemas.microsoft.com/office/word/2010/wordprocessingInk" xmlns:wne="http://schemas.microsoft.com/office/word/2006/wordml" xmlns:wps="http://schemas.microsoft.com/office/word/2010/wordprocessingShape" mc:Ignorable="w14 w15 w16se w16cid wp14">
  <w:body>
    <w:p w14:paraId="52DAE836" w14:textId="35F10B17" w:rsidR="00A954E6" w:rsidRPr="00537D95" w:rsidRDefault="00A954E6" w:rsidP="00A954E6">
      <w:bookmarkStart w:id="0" w:name="_GoBack"/>
      <w:r w:rsidRPr="00537D95">
        <w:t xml:space="preserve">no problem dude</w:t>
      </w:r>
      <w:r>
        <w:t xml:space="preserve"></w:t>
      </w:r>
      <w:r w:rsidRPr="00537D95">
        <w:t xml:space="preserve"></w:t>
      </w:r>
    </w:p>
    <w:bookmarkEnd w:id="0"/>
    <w:p w14:paraId="35401385" w14:textId="7B1795CE" w:rsidR="00413949" w:rsidRPr="00A954E6" w:rsidRDefault="00413949" w:rsidP="00A954E6"/>
    <w:sectPr w:rsidR="00413949" w:rsidRPr="00A954E6">
      <w:pgSz w:w="12240" w:h="15840"/>
      <w:pgMar w:top="1417" w:right="1701" w:bottom="1417" w:left="1701" w:header="708" w:footer="708" w:gutter="0"/>
      <w:cols w:space="708"/>
      <w:docGrid w:linePitch="360"/>
    </w:sectPr>
  </w:body>
</w:document>"
`;

exports[`sandbox custom ErrorHandler properly handles InvalidCommandError 1`] = `
"<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<w:document xmlns:wpc="http://schemas.microsoft.com/office/word/2010/wordprocessingCanvas" xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006" xmlns:o="urn:schemas-microsoft-com:office:office" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships" xmlns:m="http://schemas.openxmlformats.org/officeDocument/2006/math" xmlns:v="urn:schemas-microsoft-com:vml" xmlns:wp14="http://schemas.microsoft.com/office/word/2010/wordprocessingDrawing" xmlns:wp="http://schemas.openxmlformats.org/drawingml/2006/wordprocessingDrawing" xmlns:w10="urn:schemas-microsoft-com:office:word" xmlns:w="http://schemas.openxmlformats.org/wordprocessingml/2006/main" xmlns:w14="http://schemas.microsoft.com/office/word/2010/wordml" xmlns:w15="http://schemas.microsoft.com/office/word/2012/wordml" xmlns:wpg="http://schemas.microsoft.com/office/word/2010/wordprocessingGroup" xmlns:wpi="http://schemas.microsoft.com/office/word/2010/wordprocessingInk" xmlns:wne="http://schemas.microsoft.com/office/word/2006/wordml" xmlns:wps="http://schemas.microsoft.com/office/word/2010/wordprocessingShape" mc:Ignorable="w14 w15 wp14">
  <w:body>
    <w:p w:rsidR="00A17023" w:rsidRPr="00537D95" w:rsidRDefault="00A17023">
      <w:r w:rsidRPr="00537D95">
        <w:t xml:space="preserve">Here is the query…</w:t>
      </w:r>
    </w:p>
    <w:p w:rsidR="00413949" w:rsidRDefault="00413949">
      <w:r w:rsidRPr="0070400C">
        <w:t xml:space="preserve">Invalid command…</w:t>
      </w:r>
    </w:p>
    <w:p w:rsidR="00537D95" w:rsidRPr="0070400C" w:rsidRDefault="00537D95">
      <w:r>
        <w:t xml:space="preserve">SyntaxError: Unexpected identifier 'foo'</w:t>
      </w:r>
    </w:p>
    <w:p w:rsidR="00413949" w:rsidRPr="0070400C" w:rsidRDefault="00413949">
      <w:bookmarkStart w:id="0" w:name="_GoBack"/>
      <w:bookmarkEnd w:id="0"/>
      <w:r w:rsidRPr="0070400C">
        <w:t xml:space="preserve">Post command…</w:t>
      </w:r>
    </w:p>
    <w:sectPr w:rsidR="00413949" w:rsidRPr="0070400C">
      <w:pgSz w:w="12240" w:h="15840"/>
      <w:pgMar w:top="1417" w:right="1701" w:bottom="1417" w:left="1701" w:header="708" w:footer="708" w:gutter="0"/>
      <w:cols w:space="708"/>
      <w:docGrid w:linePitch="360"/>
    </w:sectPr>
  </w:body>
</w:document>"
`;

exports[`sandbox custom ErrorHandler properly handles InvalidCommandError 2`] = `
[
  [SyntaxError: Unexpected identifier 'foo'],
]
`;

exports[`sandbox rejectNullish setting IMAGE 1`] = `"Result of command qr(\`some random string\`) is null or undefined and rejectNullish is set"`;

exports[`throw when result of INS command is an object 1`] = `"Result of command 'companies' is an object"`;

exports[`throw when result of INS command is an object 2`] = `"Result of command 'companies' is an object"`;

exports[`throw when user tries to iterate over non-array 1`] = `"Invalid FOR command (can only iterate over Array): FOR name IN companies"`;

exports[`throw when user tries to iterate over non-array 2`] = `"Invalid FOR command (can only iterate over Array): FOR name IN companies"`;
